osquery/tables/system/linux/md_tables.cpp

```diff
@@ -0,0 +1,733 @@
+/*
+ *  Copyright (c) 2014-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+#include <errno.h>
+#include <fcntl.h>
+#include <string.h>
+#include <sys/ioctl.h>
+#include <unistd.h>
+
+#include <linux/major.h>
+#include <linux/raid/md_p.h>
+#include <linux/raid/md_u.h>
+
+#include <fstream>
+#include <memory>
+
+#include <boost/algorithm/string/trim.hpp>
+
+#include <osquery/core/conversions.h>
+#include <osquery/logger.h>
+
+#include "osquery/events/linux/udev.h"
+#include "osquery/tables/system/linux/md_tables.h"
+
+namespace osquery {
+namespace tables {
+
+std::string kMDStatPath = "/proc/mdstat";
+
+/**
+ * @brief Removes prefixing and suffixing character from each string in vector
+ *
+ * @param strs reference to vector of target strings
+ * @param c character to remove
+ *
+ */
+void trimStrs(std::vector<std::string>& strs) {
+  for (auto& s : strs) {
+    boost::algorithm::trim(s);
+  }
+}
+
+/**
+ * @brief reads mdstat file and parses as vector of lines
+ *
+ * @param lines reference to vector of strings to store lines in
+ *
+ * Trims all white space in lines and ignores empty lines (i.e. only spaces,
+ * tabs, etc.)
+ */
+inline void getLines(std::vector<std::string>& lines) {
+  std::ifstream handle(kMDStatPath);
+
+  std::string line;
+  if (handle.is_open()) {
+    while (getline(handle, line)) {
+      boost::algorithm::trim(line);
+
+      if (line.find_first_not_of("\t\r\v ") != std::string::npos) {
+        lines.push_back(line);
+      }
+    }
+
+    handle.close();
+  }
+}
+
+/**
+ * @brief function for walking thru a udev subsystem and working on devices
+ *
+ * @param systemName the name of the sysfs subsytem to work with
+ * @param f function to execute on the subsystem; returns true if can break the
+ * device loop, false to keep walking
+ *
+ */
+void walkUdevDevices(std::string systemName,
+                     std::function<bool(udev_device* const&)> f) {
```

Comment:
No (since i'm passing in lambda literals throughout the code), but i think i can be a `std::function<bool(udev_device* const&)> const& f`